src/session_impl.cpp

```diff
@@ -811,35 +808,29 @@ namespace aux {
 
 		boost::shared_ptr<plugin> p(new session_plugin_wrapper(ext));
 
-		m_ses_extensions.push_back(p);
-		m_session_extension_features |= p->implemented_features();
+		add_ses_extension(p, false);
 	}
 
-	void session_impl::add_ses_extension(boost::shared_ptr<plugin> ext)
+	void session_impl::add_ses_extension(boost::shared_ptr<plugin> ext, bool no_wrap)
 	{
 		TORRENT_ASSERT(is_single_thread());
 		TORRENT_ASSERT_VAL(ext, ext);
 
-		m_ses_extensions.push_back(ext);
-		m_alerts.add_extension(ext);
-		ext->added(session_handle(this));
-		m_session_extension_features |= ext->implemented_features();
+		boost::uint32_t features = ext->implemented_features();
 
-		// get any DHT queries the plugin would like to handle
-		// and record them in m_extension_dht_queries for lookup
-		// later
-		dht_extensions_t dht_ext;
-		ext->register_dht_extensions(dht_ext);
-		for (dht_extensions_t::iterator e = dht_ext.begin();
-			e != dht_ext.end(); ++e)
+		m_ses_extensions[plugins_all_idx].push_back(ext);
+		if (features & plugin::optimistic_unchoke_feature)
+			m_ses_extensions[plugins_optimistic_unchoke_idx].push_back(ext);
+		if (features & plugin::tick_feature)
+			m_ses_extensions[plugins_tick_idx].push_back(ext);
+		if (features & plugin::dht_request_feature)
+			m_ses_extensions[plugins_dht_request_idx].push_back(ext);
+
+		if (no_wrap)
```

Comment:
it's not obvious to me what "no_wrap" refers to (even when looking at this code). Could you maybe add a comment?